File: source/http_client.cpp
#include "http_client.hpp"

namespace HttpClient
{
    size_t writeCallback(void *contents, size_t size, size_t nmemb, std::string *userp)
    {
        userp->append((char *)contents, size * nmemb);
        return size * nmemb;
    }

    std::string get(const std::string &url, int retries)
    {
        std::string response;
        for (int attempt = 0; attempt < retries; ++attempt)
        {
            CURL *curl = curl_easy_init();
            if (curl)
            {
                curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
                curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
                curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

                // 添加 HTTP 头部以避免缓存
                struct curl_slist *headers = nullptr;
                headers = curl_slist_append(headers, "Cache-Control: no-cache");
                headers = curl_slist_append(headers, "Pragma: no-cache");
                headers = curl_slist_append(headers, "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36");
                curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

                // 启用所有支持的编码类型
                curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "");

                // 允许重定向
                curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
                curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 3L);

                curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
                curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 5L);
                curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
                curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
                curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);

                response.clear();
                CURLcode res = curl_easy_perform(curl);

                if (res != CURLE_OK)
                {
                    curl_slist_free_all(headers);
                    curl_easy_cleanup(curl);
                    if (attempt < retries - 1)
                    {
                        std::this_thread::sleep_for(std::chrono::seconds(1));
                        continue;
                    }
                    return "";
                }

                long response_code;
                curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
                if (response_code != 200)
                {
                    curl_slist_free_all(headers);
                    curl_easy_cleanup(curl);
                    if (attempt < retries - 1)
                    {
                        std::this_thread::sleep_for(std::chrono::seconds(1));
                        continue;
                    }
                    return "";
                }

                curl_slist_free_all(headers);
                curl_easy_cleanup(curl);
                break;
            }
        }
        return response;
    }

    bool checkRedirect(const std::string &url, std::string &redirect_url)
    {
        CURL *curl = curl_easy_init();
        if (!curl)
            return false;

        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 0L);
        curl_easy_setopt(curl, CURLOPT_NOBODY, 1L);
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);

        struct curl_slist *headers = nullptr;
        headers = curl_slist_append(headers, "x-device-abis: x86_64,arm64-v8a,x86,armeabi-v7a,armeabi");
        headers = curl_slist_append(headers, "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        CURLcode res = curl_easy_perform(curl);
        bool success = false;

        if (res == CURLE_OK)
        {
            char *redirect_url_raw = nullptr;
            if (curl_easy_getinfo(curl, CURLINFO_REDIRECT_URL, &redirect_url_raw) == CURLE_OK && redirect_url_raw)
            {
                redirect_url = redirect_url_raw;
                success = true;
            }
        }

        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        return success;
    }
}

File: Src/Python/http_client.py
"""
HTTP客户端模块
"""

import time
import requests
from typing import Optional
import urllib3

from rich.console import Console

# 禁用SSL警告
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

console = Console()


class HttpClient:
    """HTTP客户端类"""

    # 默认的HTTP请求头
    DEFAULT_HEADERS = {
        "Cache-Control": "no-cache",
        "Pragma": "no-cache",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    }

    @staticmethod
    def get(url: str, retries: int = 5, timeout: int = 10) -> str:
        """
        执行HTTP GET请求并返回响应内容

        Args:
            url: 请求的URL
            retries: 重试次数，默认为5次
            timeout: 超时时间（秒），默认为10秒

        Returns:
            服务器响应的内容，如果请求失败则返回空字符串
        """
        for attempt in range(retries):
            try:
                response = requests.get(
                    url,
                    headers=HttpClient.DEFAULT_HEADERS,
                    timeout=timeout,
                    verify=False,  # 不验证SSL证书
                    allow_redirects=True,
                )

                if response.status_code == 200:
                    return response.text

                # 如果状态码不是200，则继续重试
                if attempt < retries - 1:
                    time.sleep(1)
                    continue

                return ""
            except requests.exceptions.RequestException:
                if attempt < retries - 1:
                    time.sleep(1)
                    continue
                return ""

        return ""

    @staticmethod
    def check_redirect(url: str) -> tuple[bool, str]:
        """
        检查URL是否有效（不执行实际请求，只检查重定向）

        Args:
            url: 要检查的URL

        Returns:
            (是否有效, 重定向URL)的元组
        """
        try:
            headers = {
                **HttpClient.DEFAULT_HEADERS,
                "x-device-abis": "x86_64,arm64-v8a,x86,armeabi-v7a,armeabi",
            }

            response = requests.head(
                url, headers=headers, allow_redirects=False, verify=False, timeout=5
            )

            redirect_url = response.headers.get("Location", "")
            if redirect_url:
                return True, redirect_url

            return False, ""
        except Exception:
            return False, ""

    @staticmethod
    def get_json(url: str, retries: int = 5) -> Optional[dict]:
        """
        执行HTTP GET请求并返回JSON响应

        Args:
            url: 请求的URL
            retries: 重试次数，默认为5次

        Returns:
            解析后的JSON对象，失败时返回None
        """
        response_text = HttpClient.get(url, retries)
        if not response_text:
            return None

        try:
            import json

            return json.loads(response_text)
        except json.JSONDecodeError:
            return None
